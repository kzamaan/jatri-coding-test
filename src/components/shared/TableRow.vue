<script setup>
    import PrimaryButton from '@/components/PrimaryButton.vue';
    import { useProducts } from '@/composable/useProducts';
    import { ref, watch } from 'vue';
    import ErrorMessage from './ErrorMessage.vue';
    import ProductDetail from './ProductDetail.vue';

    const { fetchProductById, productDetail, isLoading, error } = useProducts();

    const props = defineProps({
        product: {
            type: Object,
            default: () => ({}),
        },
        skipped: {
            type: Number,
            default: 0,
        },
        iteration: {
            type: Number,
            default: 0,
        },
    });

    watch(productDetail, value => {
        console.log('Product detail', value);
    });

    const isShowDetail = ref(false);

    const showProductDetail = () => {
        isShowDetail.value = !isShowDetail.value;
        if (!productDetail.value.id) {
            fetchProductById(props.product.id);
        }
    };
</script>

<template>
    <tr class="bg-white border-b dark:bg-gray-800 dark:border-gray-700">
        <th scope="row" class="px-6 py-4 border font-medium text-gray-900 whitespace-nowrap dark:text-white">
            {{ skipped + (iteration + 1) }}
        </th>
        <td class="px-6 py-4 border">{{ product.title }}</td>
        <td class="px-6 py-4 border">{{ product.rating }}</td>
        <td class="px-6 py-4 border text-right">{{ Math.round(product.price).toLocaleString() }}</td>
        <td class="px-6 py-4 border text-center">
            <PrimaryButton @click="showProductDetail" class="w-20">
                {{ isLoading ? 'Loading...' : isShowDetail ? 'Close' : 'Show' }}
            </PrimaryButton>
        </td>
    </tr>
    <template v-if="isShowDetail">
        <ProductDetail :product-detail="productDetail" :is-loading="isLoading" />
        <error-message :errorMessage="error" />
    </template>
</template>
