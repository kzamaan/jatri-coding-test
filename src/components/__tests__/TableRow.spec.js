import { describe, expect, it } from 'vitest';

import { mount } from '@vue/test-utils';
import TableRow from '../shared/TableRow.vue';

describe('TableRow', () => {
    it('renders table row properly', () => {
        const wrapper = mount(TableRow, {
            props: {
                product: {
                    id: 1,
                    title: 'iPhone 9',
                    description: 'An apple mobile which is nothing like apple',
                    price: 549,
                    discountPercentage: 12.96,
                    rating: 4.69,
                },
            },
        });
        expect(wrapper.html()).toMatchSnapshot();
    });
});
