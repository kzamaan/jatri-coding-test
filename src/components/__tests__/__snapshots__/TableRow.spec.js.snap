// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableRow > renders table row properly 1`] = `
"<tr class=\\"bg-white border-b dark:bg-gray-800 dark:border-gray-700\\">
  <th scope=\\"row\\" class=\\"px-6 py-4 border font-medium text-gray-900 whitespace-nowrap dark:text-white\\">1</th>
  <td class=\\"px-6 py-4 border\\">iPhone 9</td>
  <td class=\\"px-6 py-4 border\\">4.69</td>
  <td class=\\"px-6 py-4 border text-right\\">549</td>
  <td class=\\"px-6 py-4 border text-center\\"><button type=\\"button\\" class=\\"btn w-20\\">Show</button></td>
</tr>
<!--v-if-->"
`;
